cluster-autoscaler/cloudprovider/clusterapi/clusterapi_controller.go

```diff
@@ -639,80 +574,109 @@ func (c *machineController) findNodeByProviderID(providerID normalizedProviderID
 	return node.DeepCopy(), nil
 }
 
-func (c *machineController) getMachine(namespace, name string, options metav1.GetOptions) (*Machine, error) {
-	u, err := c.dynamicclient.Resource(*c.machineResource).Namespace(namespace).Get(context.TODO(), name, options)
-	if err != nil {
-		return nil, err
+func (c *machineController) listMachinesForScalableResource(r *unstructured.Unstructured) ([]*unstructured.Unstructured, error) {
+	switch r.GetKind() {
+	case machineSetKind, machineDeploymentKind:
+		unstructuredSelector, found, err := unstructured.NestedMap(r.UnstructuredContent(), "spec", "selector")
+		if err != nil {
+			return nil, err
+		}
+
+		if !found {
+			return nil, fmt.Errorf("expected field spec.selector on scalable resource type")
+		}
+
+		labelSelector := &metav1.LabelSelector{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstructuredSelector, labelSelector); err != nil {
+			return nil, err
+		}
+
+		selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+		if err != nil {
+			return nil, err
+		}
+
+		return listResources(c.machineInformer.Lister().ByNamespace(r.GetNamespace()), clusterNameFromResource(r), selector)
+	default:
+		return nil, fmt.Errorf("unknown scalable resource kind %s", r.GetKind())
 	}
-	return newMachineFromUnstructured(u.DeepCopy()), nil
 }
 
-func (c *machineController) getMachineSet(namespace, name string, options metav1.GetOptions) (*MachineSet, error) {
-	u, err := c.dynamicclient.Resource(*c.machineSetResource).Namespace(namespace).Get(context.TODO(), name, options)
+func (c *machineController) listScalableResources() ([]*unstructured.Unstructured, error) {
+	scalableResources, err := c.listResources(c.machineSetInformer.Lister())
 	if err != nil {
 		return nil, err
 	}
-	return newMachineSetFromUnstructured(u.DeepCopy()), nil
-}
 
-func (c *machineController) getMachineDeployment(namespace, name string, options metav1.GetOptions) (*MachineDeployment, error) {
-	u, err := c.dynamicclient.Resource(*c.machineDeploymentResource).Namespace(namespace).Get(context.TODO(), name, options)
-	if err != nil {
-		return nil, err
+	if c.machineDeploymentsAvailable {
+		machineDeployments, err := c.listResources(c.machineDeploymentInformer.Lister())
+		if err != nil {
+			return nil, err
+		}
+
+		scalableResources = append(scalableResources, machineDeployments...)
 	}
-	return newMachineDeploymentFromUnstructured(u.DeepCopy()), nil
+	return scalableResources, nil
 }
 
-func (c *machineController) listMachines(namespace string, selector labels.Selector) ([]*Machine, error) {
-	objs, err := c.machineInformer.Lister().ByNamespace(namespace).List(selector)
-	if err != nil {
-		return nil, err
+func (c *machineController) listResources(lister cache.GenericLister) ([]*unstructured.Unstructured, error) {
```

Comment:
Need to add test cases for this method to verify that autodiscovery config is doing the right thing.